addons/web_editor/static/lib/odoo-editor/src/utils/utils.js

```diff
@@ -755,6 +755,252 @@ export function preserveCursor(document) {
     };
 }
 
+//------------------------------------------------------------------------------
+// Format utils
+//------------------------------------------------------------------------------
+
+const formatsInfos = {
+    italic: {
+        tagName: 'em',
+        isFormatted: isItalic,
+        isTag: (node) => ['EM', 'I'].includes(node.tagName),
+        hasStyle: (node) => Boolean(node.style && node.style['font-style']),
+        addStyle: (node) => node.style['font-style'] = 'italic',
+        addNeutralStyle: (node) => node.style['font-style'] = 'normal',
+        removeStyle: (node) => removeStyle(node, 'font-style'),
+    },
+    bold: {
+        tagName: 'strong',
+        isFormatted: isBold,
+        isTag: (node) => ['STRONG', 'B'].includes(node.tagName),
+        hasStyle: (node) => Boolean(node.style && node.style['font-weight']),
+        addStyle: (node) => node.style['font-weight'] = 'bolder',
+        addNeutralStyle: (node) => {
+            node.style['font-weight'] = 'normal'
+        },
+        removeStyle: (node) => removeStyle(node, 'font-weight'),
+    },
+    underline: {
+        tagName: 'u',
+        isFormatted: isUnderline,
+        isTag: (node) => node.tagName === 'U',
+        hasStyle: (node) => node.style && node.style['text-decoration-line'].includes('underline'),
+        addStyle: (node) => addStyleItem(node, 'text-decoration-line', 'underline'),
+        removeStyle: (node) => removeStyleItem(node, 'text-decoration-line', 'underline'),
+    },
+    strikeThrough: {
+        tagName: 's',
+        isFormatted: isStrikeThrough,
+        isTag: (node) => node.tagName === 'S',
+        hasStyle: (node) => node.style && node.style['text-decoration-line'].includes('line-through'),
+        addStyle: (node) => addStyleItem(node, 'text-decoration-line', 'line-through'),
+        removeStyle: (node) => removeStyleItem(node, 'text-decoration-line', 'line-through'),
+    },
+    fontSize: {
+        isFormatted: isFontSize,
+        hasStyle: (node, props) => node.style && node.style['font-size'] === props.size,
+        addStyle: (node, props) => node.style['font-size'] = props.size,
+        removeStyle: (node) => removeStyle(node, 'font-size'),
+    },
+    switchDirection: {
+        isFormatted: isDirectionSwitched,
+    }
+
+}
+
+const isBlockOrHasClass = (node) => {
+    return isBlock(node) || (node.classList && Boolean(node.classList.length));
+}
+const ancestorsUntilBlockOrHasClass = (node) => {
```

Comment:
This is used only once outside of its own recursivity. Couldn't it be replaced by a loop at the call site ?

Reply:
I find the loop more cubersome and more difficult to understand:
```js
        const ancestorsUntilBlockOrHasClass = [];
        let node = selectedTextNode.parentElement;
        while (node && !isBlock(node) && !(node.classList && Boolean(node.classList.length))) {
            ancestorsUntilBlockOrHasClass.push(node);
            node = node.parentElement;
        }
        ancestorsUntilBlockOrHasClass.push(node);
        const [blockOrInlineClass, ...inlineAncestors] = ancestorsUntilBlockOrHasClass.reverse();
```

Reply:
You're saying that this :
```js
const isBlockOrHasClass = (node) => {
    return isBlock(node) || (node.classList && Boolean(node.classList.length));
}
const ancestorsUntilBlockOrHasClass = (node) => {
    if (!node || !node.parentElement || isBlockOrHasClass(node)) return [];
    return [node.parentElement, ...ancestorsUntilBlockOrHasClass(node.parentElement)];
};
const [blockOrInlineClass, ...inlineAncestors] = ancestorsUntilBlockOrHasClass(selectedTextNode).reverse();
```
Is better than this :
```js
// Compute inline ancestors until finding one which is a block or has a class.
const ancestors = [];
let node = selectedTextNode.parentElement;
while (node && !isBlock(node) && !(node.classList && node.classList.length)) { 
    ancestors.push(node);
    node = node.parentElement;
} 
ancestors.push(node);
const [blockOrInlineClass, ...inlineAncestors] = ancestors.reverse();
```
I already disagree at that point but, even if I did agree, I would argue that it's because you didn't push the conversion from one version to the other far enough. I'd write it like this:
```js
// Compute inline ancestors until finding one which is a block or has a class.
let blockOrInlineClass;
const inlineAncestors = [];
let node = selectedTextNode;
while (!blockOrInlineClass && (node = node.parentElement)) {
    if (isBlock(node) || (node.classList && node.classList.length)) {
        blockOrInlineClass = node;
    } else {
        inlineAncestors.unshift(node);
    }
} 
```
This code says _exactly_ what it does at each point, does not write `node.parentElement` multiple times, never returns deliberately empty values, does not construct the array in the wrong order just to call reverse at the end, and does not return a tuple. (Because yes, pushing the blockOrInlineClass ancestor in the same array as the other ancestors is somewhat hiding a tuple return type, and the proof is that you have to deconstruct the return value to separate the two different kinds of result returned by the function.)

I guess you'll disagree that it is better than the version with the ultra ad-hoc helpers so what do we do now ? ^^'

Reply:
Suggestion applied.